### PR TITLE
Application Server PubSub events

### LINKED ISSUE #1017: Flaky AS NATS test
#### Summary

Flaky AS NATS test

#### Steps to Reproduce

Random

#### What do you see now?

```
  WARN No cookie hash key configured, generated a random one hash_key=CA90E1F03FA690E96351D8FB1163251BE7205552608AD070C6297727C322A7C100EF2DCD28DDF676B5B4B4529159585A748B4CC14F01F2685C4277897AF3E535 namespace=web test_name=TestApplicationServer
  WARN No cookie block key configured, generated a random one block_key=B2E6930B986F933271CB083C3912A2A2B056AE91196DAB0FA69E46DED7F24E39 namespace=web test_name=TestApplicationServer
  WARN No static assets found in any search path namespace=web search_paths=[] test_name=TestApplicationServer
  WARN No rights TTL configured                 test_name=TestApplicationServer
  WARN No cluster key configured, generated a random one key=58faddf0ecadac5f9f637c0e088435f5f8466f7fc1a0a89edabac396815ebd7c test_name=TestApplicationServer
  INFO Listening for connections                address=:9184 namespace=grpc protocol=gRPC test_name=TestApplicationServer
  INFO Listening for connections                address=:8099 namespace=web protocol=Web test_name=TestApplicationServer
  INFO Linked                                   application_uid=foo-app namespace=applicationserver network_server=ns test_name=TestApplicationServer
  INFO Subscribed                               application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  INFO Unsubscribed                             application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  WARN Finished streaming call                  duration=2.048457872s error=context canceled grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7VXZEKX3GKJ8D3AH9Q14 test_name=TestApplicationServer
  WARN Finished streaming call                  duration=932.957µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=a5a973c3896743c9bb59988d03de8fab error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY0S0Z15Z5CRB8GFYXF test_name=TestApplicationServer
  WARN Finished streaming call                  duration=541.678µs error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) error_correlation_id=210204d477db4da882c6f747891e3f87 error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY11FX41KG7PCHMSYME test_name=TestApplicationServer
  INFO Subscribed                               application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  WARN Drop downlink message; session not found or AppSKey not available application_uid=foo-app count=3 device_uid=foo-app.foo-device f_cnt=12 f_port=12 namespace=applicationserver network_server=ns new_session_key_id=44 next_a_f_cnt_down=85 session_key_id=11223344 test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.064318ms error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) error_correlation_id=584f9b7bc98f4325b9846e4e36a33918 error_name=device_not_found error_namespace=pkg/applicationserver grpc_code=NotFound grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K87ZGJS24S3G86BA4K032 test_name=TestApplicationServer
  INFO Finished unary call                      duration=2.152703ms grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K89ZHJVW15MYJ2CEES28C test_name=TestApplicationServer
  INFO Finished unary call                      duration=2.944889ms grpc_method=DownlinkQueuePush grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K8BZHJW3QHJ8K8QPRDRH0 test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.453864ms grpc_method=DownlinkQueueReplace grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K8DZJ87CJ6T5KTHJ6PX0R test_name=TestApplicationServer
  INFO Unsubscribed                             application_uid=foo-app grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=applicationserver/io/grpc remote_addr=pipe request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  WARN Finished streaming call                  duration=18.48105067s error=context canceled grpc_method=Subscribe grpc_service=ttn.lorawan.v3.AppAs namespace=grpc request_id=01DG4K7XY2QBQ42837ZSTQFNRA test_name=TestApplicationServer
  INFO Connected                                application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35342 test_name=TestApplicationServer
  INFO Disconnected                             application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35342 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35346 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35348 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35352 test_name=TestApplicationServer
  WARN Failed to setup connection               error=error:pkg/auth/rights:no_application_rights (no rights for application `invalid-application`) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35354 test_name=TestApplicationServer
  INFO Connected                                application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer
  WARN Drop downlink message; session not found or AppSKey not available application_uid=foo-app count=3 device_uid=foo-app.foo-device f_cnt=12 f_port=12 namespace=applicationserver network_server=ns new_session_key_id=44 next_a_f_cnt_down=85 session_key_id=11223344 test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to process upstream message       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver network_server=ns test_name=TestApplicationServer
  WARN Failed to handle downlink messages       application_uid=foo-app error=error:pkg/applicationserver:device_not_found (device `foo-app.bar-device` not found) namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer topic=v3/foo-app/devices/bar-device/down/push
  INFO Disconnected                             application_uid=foo-app namespace=applicationserver/io/mqtt remote_addr=127.0.0.1:35358 test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.400271ms grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K98147PW65SEEHSF5JF21 test_name=TestApplicationServer
  INFO Started                                  application_uid=foo-app namespace=applicationserver/io/pubsub provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  WARN Finished unary call                      duration=751.208µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=b626925a3ee64fce8c140b233406371f error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A1444G0TT28MHP8WFBR test_name=TestApplicationServer
  WARN Finished unary call                      duration=505.353µs error=error:pkg/auth/rights:no_application_rights (no rights for application `foo-app`) error_correlation_id=894aaf0d399a49be8345bb48362e332f error_name=no_application_rights error_namespace=pkg/auth/rights grpc_code=PermissionDenied grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A164E0JST222EMR2A7Y test_name=TestApplicationServer
  WARN Failed to receive downlink queue operation application_uid=foo-app error=context canceled namespace=applicationserver/io/pubsub operation=push provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Finished unary call                      duration=1.562081ms grpc_method=Set grpc_service=ttn.lorawan.v3.ApplicationPubSubRegistry namespace=grpc request_id=01DG4K9A17JAWTQEA2AEDZ18QX test_name=TestApplicationServer
  WARN Failed to receive downlink queue operation application_uid=foo-app error=context canceled namespace=applicationserver/io/pubsub operation=replace provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Integration canceled                     application_uid=foo-app namespace=applicationserver/io/pubsub provider=*ttnpb.ApplicationPubSub_NATS pub_sub_id=test test_name=TestApplicationServer
  INFO Unlinked                                 application_uid=foo-app namespace=applicationserver network_server=ns test_name=TestApplicationServer
--- FAIL: TestApplicationServer (55.51s)
    --- FAIL: TestApplicationServer/Traffic/pubsub/nats (4.10s)
        --- FAIL: TestApplicationServer/Traffic/pubsub/nats/Upstream (2.05s)
            --- FAIL: TestApplicationServer/Traffic/pubsub/nats/Upstream/RegisteredDevice/JoinAccept (2.05s)
                applicationserver_test.go:1504: Expected upstream timeout
FAIL
FAIL	go.thethings.network/lorawan-stack/pkg/applicationserver	55.576s
```

#### What do you want to see instead?

No errors

#### Environment

Travis, i386

#### How do you propose to implement this?

Probably increase timeouts. AS is already taking a lot of time though.

#### Can you do this yourself and submit a Pull Request?

Can review

### LINKED ISSUE #1286: Add events to the PubSub integrations frontend
#### Summary

Add events to the PubSub integrations frontend.

#### Why do we need this?

In order to be able to follow asynchronous actions that occur in the PubSub frontend. Also this is one way of achieving test determinism for #1017.

#### What is already there? What do you see now?

No events.

#### What is missing? What do you want to see?

Outside of the classical CRUD events, the events for started (so fully connected both upstream and downstream), stopped and failed would be very helpful.

#### How do you propose to implement this?

Add CRUD events and the start, stop and failure events.

#### Can you do this yourself and submit a Pull Request?

Yes.

### PR DESCRIPTION
**Pull request by adriansmares**
#### Summary

Closes #1286 
Closes #1017 
#### Changes

- Add set, delete, start, stop and failed events to the PubSub integrations.
- Expand `pkg/util/test` with event redirection / waiting capabilities
- Cut the AS global test timeout in half.
- Wait for integrations to start before sending / receiving commands from them in tests.
- Ensure that integrations and links close properly.

#### Notes for Reviewers

e4454060fef496fe5aaf4d2bb182e747be72cd39 and 1b80ac3c9b91a39a6e0c1c0d2eed30f4e237c170 are two important commits since they fix a race condition in the linking / integration process. Calling `stop` on an integration / link is expected to fully close the integration / link, such that a new `start` call can occur without an `already started` error. However, the behavior before these two commits was that `stop` will simply cancel the context. This is not correct, since the goroutine that clears the integration / link from the `sync.Map` that holds the integrations / links does not have to be scheduled before we start the linking / integration task. This is very important since it can result in `Set` calls that do not start the integration / link, if they were already started.

~Does moving `ApplicationPubSubIdentifiers` from one proto file to another count as breaking API ?~

#### Release Notes

- Added Application Server PubSub integrations events ! See `pkg/applicationserver/io/pubsub/observability.go` for more details.

### CODE REVIEW

File: pkg/applicationserver/io/pubsub/observability.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	"go.thethings.network/lorawan-stack/pkg/events"
+	"go.thethings.network/lorawan-stack/pkg/metrics"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+)
+
+var (
+	evtSetPubSub = events.Define(
+		"as.pubsub.set", "set pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_READ,
```
Comment (htdvisser) [inline pkg/applicationserver/io/pubsub/observability.go]: I think these rights should only have `SETTINGS`, (not `TRAFFIC`). Keep in mind that with events it's OR (with `rights.Require` in the registry it's AND)

File: pkg/applicationserver/io/pubsub/observability.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE,
+	)
+	evtDeletePubSub = events.Define(
+		"as.pubsub.delete", "delete pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_READ,
```
Comment (htdvisser) [inline pkg/applicationserver/io/pubsub/observability.go]: Also only `SETTINGS`

File: pkg/applicationserver/io/pubsub/observability.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+		ttnpb.RIGHT_APPLICATION_TRAFFIC_DOWN_WRITE,
+	)
+	evtPubSubStart = events.Define(
+		"as.pubsub.start", "start pubsub",
+		ttnpb.RIGHT_APPLICATION_SETTINGS_BASIC,
```
Comment (htdvisser) [inline pkg/applicationserver/io/pubsub/observability.go]: Here and below probably also the `TRAFFIC` rights.

**Comment by adriansmares** [comment]
> Mostly looks good, but I have my concerns about the identifiers.

An alternative is to publish these events using the application identifiers, and the PubSub identifiers as part of the body. Do keep in mind however that this becomes a bit `weird`: take the failed integration event - right now it contains only the error as data, and the events package detects this automatically at the top level and handles it correctly. If we use the application identifiers - we have to submit both the PubSub identifiers and the error - this becomes tricky since the error is no longer marshaled correctly.

I'm more than open to suggestions about how we can handle this in an uniform way.

File: pkg/applicationserver/applicationserver_util_test.go [hunk 4]
```diff
@@ -22,7 +22,8 @@ import (
 )
 
 var (
-	Timeout = (1 << 9) * test.Delay
+	Timeout          = (1 << 8) * test.Delay
```
Comment (johanstokking) [inline pkg/applicationserver/applicationserver_util_test.go]: Note that this is still 4 seconds with `TEST_SLOWDOWN=16`

Can we get this lower and add `Timeout * X` in particular places that are notoriously slow?

Comment (adriansmares) [reply pkg/applicationserver/applicationserver_util_test.go]: Travis runs on `TEST_SLOWDOWN=8`. With that being said, I can try to squeeze it to `(1 << 7)` and double it on a case-by-case basis.

**Pull request merged**